strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -436,49 +434,113 @@ def measure_dyne(self, covmat, indices, shots=1):
         if covmat.shape != (2 * len(indices), 2 * len(indices)):
             raise ValueError("Covariance matrix size does not match indices provided")
 
+        if np.linalg.det(covmat) < (self.hbar / 2) ** (2 * len(indices)):
+            raise ValueError("Measurement covariance matrix is unphysical.")
+
+        if self.covs.imag.any():
+            raise NotImplementedError("Covariance matrices must be real")
+
         for i in indices:
             if self.active[i] is None:
                 raise ValueError("Cannot apply measurement, mode does not exist")
 
-        print("Measurements implemened for real-valued Gaussians only.")
-
         expind = np.concatenate((2 * np.array(indices), 2 * np.array(indices) + 1))
         vals = np.zeros((shots, 2 * len(indices)))
-        pos_weights_ind = np.where(np.array(self.weights).real > 0)[0]
-        pos_weights = np.array(self.weights)[pos_weights_ind].real
-        pos_weights = pos_weights / np.sum(pos_weights)
+        imag_means_ind = np.where(self.means[:, expind].imag.any(axis=1))[0]
+        nonneg_weights_ind = np.where(np.angle(self.weights) != np.pi)[0]
```

Comment:
Let's just keep this for the time being since it works. We're not exactly trying to find nonzero elements. We want elements that don't have negative weights and real means.